[RFC] Feature #26 export to ical

NMFR opened the pull request.

NMFR committed:
Adopt Go 1.11 Modules as the dependency manager

NMFR committed:
Add export command and ical exporter

To export run:
```sh
golog export
```

Or:
```sh
golog export ical golog.ics
```

Closes #26

NMFR committed:
Change travis CI tests versions

Version 1.1.2 is being removed due to and error on the github.com/codegangsta/cli@v1.20.0 dependency:
````
../../codegangsta/cli/context.go:170: undefined: flag.Getter
```

Seems codegangsta/cli is using go's flag.Getter that was only [added in go v1.2](https://github.com/golang/go/commit/49b3301f4ce2a9e21ac076ed53d4bd6e775e748d)

Other versions were replaced by 1.11.4 because the project started using go modules

mlimaloureiro commented on golog.go:
```diff
@@ -45,6 +51,13 @@ var commands = []cli.Command{
+	{
+		Name:   		"export",
+		Usage:  		"Export all tasks in a specific format",
+		ArgsUsage:   	fmt.Sprintf("[%s] [filePath]", strings.Join(ExporterNames, " | ")),
```
run go fmt

mlimaloureiro commented on golog.go:
```diff
@@ -45,6 +51,13 @@ var commands = []cli.Command{
+		Usage:  		"Export all tasks in a specific format",
+		ArgsUsage:   	fmt.Sprintf("[%s] [filePath]", strings.Join(ExporterNames, " | ")),
+		Action: 		Export,
+		BashComplete: 	AutocompleteExport,
```
go fmt

mlimaloureiro commented on ical_exporter.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func (iCalTaskExporter ICalTaskExporter) getPRODID() string {
+	if iCalTaskExporter.PRODID == "" {
```
what's the PRODID? i'm missing context from ical

NMFR replied:
PRODID: This property specifies the identifier for the product that created the iCalendar object.
Needs to be unique and seems to be mandatory. 

https://www.kanzaki.com/docs/ical/prodid.html

mlimaloureiro commented on ical_exporter.go:
```diff
@@ -0,0 +1,129 @@
+
+func (iCalTaskExporter ICalTaskExporter) getPRODID() string {
+	if iCalTaskExporter.PRODID == "" {
+		return "-//mlimaloureiro/golog"
```
should this be parameterized ?

NMFR replied:
Yeah you are right, should be a constant.

mlimaloureiro commented on ical_exporter.go:
```diff
@@ -0,0 +1,129 @@
+	var error error
+
+	tryWriteStrings(&error, writer, "BEGIN:VCALENDAR\n")
+	if error == nil {
```
```
var err error
if err == nil {
```

let's use vertical proximity to check where to define the variables

mlimaloureiro replied:
also, this is a nice topic. for defining errors, I think we should standardize `err` for the variable to avoid confusion with the `error` keyword. what do you think ? need input here aguiar

NMFR replied:
I thought about using `err` instead of `error` but was afraid it would lose readability.

What do you mean with "vertical proximity"? is it about the new line @ line 55?

jsaguiar replied:
we should use `err`. if you do like that, you don't have access to error interface (error is an interface) in the function again. 

please refactor to use `err` instead. As Miguel said it's the go standard.

jsaguiar commented:
das

mlimaloureiro commented on ical_exporter.go:
```diff
@@ -0,0 +1,129 @@
+
+import "io" 
+
+type ICalTaskExporter struct {
```
I think we should think and discuss a bit more the design. We have `repositories` and `exporters` which may be a bit ambiguous. if we think very high level, we are logging time and outputting it somewhere, some questions are:

- do we want only one type of output / persistence ?
- do we want to consider an output as a way of persistence ? imagine we configure it to just save in ical, this means it should implement a `repository` interface not an `exporter` interface
- do we want to persist in several repositories ? if so we would need a registry. how are we thinking of this ?

These just came to my mind very quickly, more questions will probably come. maybe we should think a bit about how we want to look at our problem and design our code accordingly because currently it feels `repository` and `exporter` interface kind of do the same

NMFR replied:
I didn't invest much time thinking about this yet. 
My logic here was that i didn't want to implement a parser for the ical file format. 

The current implementation of `RepositoryInterface` is more of a importer exporter where it knows how to read and write `Tasks` into some format. 
I predict this will change either to a CRUD like interface or more focused on the operations around tasks (stop task X, edit task activity start date, ...). 

I was seeing the app using a single repository only.

mlimaloureiro commented on ical_exporter_test.go:
```diff
@@ -0,0 +1,92 @@
+		t.Error("Failed export.")
+	}
+
+	expectedResult := `BEGIN:VCALENDAR
```
would put this into a fixtures file with constants or in the same file, but not in the body of the function

NMFR replied:
I am a fan of placing text near the comparison in tests unless it is a big CLOB. But this is more of habit i have when debugging in an IDE so i can quickly check the differences. But if i used a testing lib function that highlights the differences this would lose value.

I ll move them to constants in the top of the file.

jsaguiar commented on golog.go:
```diff
@@ -4,12 +4,19 @@ import (
+	ExportICal = "ical"
+)
+
+var ExporterNames = []string{ExportICal}
```
why do you need this as global variable?

NMFR replied:
I don't need it as a global variable, but i was seeing it as a constant (array of exporters the CLI uses).

What's a better alternative? Making a getExporterNames func?

jsaguiar replied:
you can take advantage of higher-order functions and pass it as parameter

jsaguiar commented on ical_exporter.go:
```diff
@@ -0,0 +1,123 @@
+// tryWriteStrings is a helper function used to write strings to a writer if an error has not previously occured
+// 	if an error occurs during the write the "error" argument will receive the error
+func tryWriteStrings(error *error, writer io.Writer, strings ...string) {
+	if strings != nil {
```
for readability proposes use early return.
```go 
if strings == nil { return }
```

jsaguiar commented on golog.go:
```diff
@@ -4,12 +4,19 @@ import (
 )
 
+const (
+	ExportICal = "ical"
```
missing comment. All exported const should have a comment, (do we really need it exported? )

jsaguiar commented on ical_exporter.go:
```diff
@@ -0,0 +1,123 @@
+	return nil
+}
+
+func (iCalTaskExporter ICalTaskExporter) GetFileExtension() string {
```
missing comment. 
Run golint to check for this

jsaguiar commented on golog.go:
```diff
@@ -4,12 +4,19 @@ import (
+	ExportICal = "ical"
+)
+
+var ExporterNames = []string{ExportICal}
```
comment or unexport

jsaguiar commented on ical_exporter.go:
```diff
@@ -0,0 +1,123 @@
+
+// tryWriteStrings is a helper function used to write strings to a writer if an error has not previously occured
+// 	if an error occurs during the write the "error" argument will receive the error
+func tryWriteStrings(error *error, writer io.Writer, strings ...string) {
```
this is a go anti pattern. You should deal with errors ASAP. 
Also new errors should be returned by the function and not mutating the previous one.

I would suggest you to take advantage of panic recover or use higher-order functions.

NMFR replied:
My idea here was to make the code more readable and avoid repeating the pattern:
``` go
if err := writeString(writer, "BEGIN:VCALENDAR\n"); err != nil {
    return err
}

```
for every string that needed to be written to the stream. When writing many strings the code gets harder to read.

I ll change it to something like this:
``` go
if err := writeStrings(
	writer,
	"BEGIN:VEVENT\n",
	"SUMMARY:", startTask.getIdentifier(), "\n",
	"DTSTART:", startTime.UTC().Format(iCalTaskExporter.getTimeFormat()), "\n",
); err != nil {
	return err
}
```

jsaguiar replied:
I understood what you try to accomplish that's why I made those suggestions.

jsaguiar commented on ical_exporter.go:
```diff
@@ -0,0 +1,123 @@
+
+import "io"
+
+const PRODID = "-//mlimaloureiro/golog"
```
comment or unexport  variable

jsaguiar reviewed (changes requested):
/Users/joao/Desktop/golog/golog.go:15: exported const ExportICal should have comment (or a comment on this block) or be unexported
/Users/joao/Desktop/golog/golog.go:18: exported var ExporterNames should have comment or be unexported
/Users/joao/Desktop/golog/ical_exporter.go:5: exported const PRODID should have comment or be unexported
/Users/joao/Desktop/golog/ical_exporter.go:7: exported type ICalTaskExporter should have comment or be unexported
/Users/joao/Desktop/golog/ical_exporter.go:83: exported method ICalTaskExporter.Export should have comment or be unexported
/Users/joao/Desktop/golog/ical_exporter.go:121: exported method ICalTaskExporter.GetFileExtension should have comment or be unexported
/Users/joao/Desktop/golog/tasks.go:4: exported const TaskStart should have comment (or a comment on this block) or be unexported

NMFR committed:
Refacture Task and Repository model structure

NMFR committed:
Refactor ICalTaskExporter into a TaskRepositoryInterface

Add SetTask and SetTasks methods to TaskRepositoryInterface
Implement the new TaskRepositoryInterface methods
Add csv export format to the CLI
Add comments to public structs / methods

NMFR committed:
Refactor models and repositories folder structure

NMFR committed:
Create a service layer for the task repository

Closes #7

mlimaloureiro commented on models/tasks/activity_test.go:
```diff
@@ -0,0 +1,47 @@
+		})
+
+		t.Run("not running", func(t *testing.T) {
+			taskActivity := TaskActivity{StartDate: timeFromString("2019-01-01T10:00:00Z"), EndDate: timeFromString("2019-01-02T10:00:00Z")}
```
we should align on max columns p/ line and how/when to break lines. usually we have either 100 or 120 max columns

mlimaloureiro replied:
also add this to our go guidelines, just saw we don't have it

NMFR replied:
https://github.com/fsschmitt/kilabs-engineering/pull/3

mlimaloureiro commented on models/tasks/activity_test.go:
```diff
@@ -0,0 +1,47 @@
+
+		t.Run("not running", func(t *testing.T) {
+			taskActivity := TaskActivity{StartDate: timeFromString("2019-01-01T10:00:01Z"), EndDate: timeFromString("2019-01-02T10:00:00Z")}
+			assert.Equal(t, 24*time.Hour-1*time.Second, taskActivity.Duration())
```
add space between operators for readability:

```
24 * time.Hour - 1 * ...
```

NMFR replied:
`go fmt` removes the spaces there.

mlimaloureiro commented on models/tasks/activity.go:
```diff
@@ -0,0 +1,23 @@
+// TaskActivity represents a date interval of a task execution
+type TaskActivity struct {
+	StartDate time.Time
+	EndDate   time.Time // EndDate.IsZero() indicates a running task, meaning the 0001-01-01 EndDate cannot be represented
```
should we have in line comments, or comments in the previous line ? let's also add this to the standards if we don't have it

NMFR replied:
If we have to pick only one i would vote for comments in the previous line. Helps keeping line length low and provides a explanation / comment before the actual code

But do we really need to pick only one? Sometimes i think it makes sense inline, like in this occurrence where it seems to make more sense to have the comment explaining the meaning of `EndDate.IsZero()` after `EndDate` has been declared.

mlimaloureiro commented on models/tasks/activity_test.go:
```diff
@@ -0,0 +1,47 @@
+
+func TestTaskActivity(t *testing.T) {
+	t.Run("IsRunning()", func(t *testing.T) {
+		t.Run("running", func(t *testing.T) {
```
we should be super explicit in the sentences we use for tests

mlimaloureiro commented on models/tasks/activity_test.go:
```diff
@@ -0,0 +1,47 @@
+			assert.Equal(t, expectedTaskDuration, taskDuration)
+		})
+
+		t.Run("not running", func(t *testing.T) {
```
_I don't quite get it why the test is called `not running` and the assertion implies that we're checking that the calc of the duration is correct_

the above would be my first comment but I spent more time understanding the test. We should be more clear in what we are trying to test

mlimaloureiro commented on models/tasks/activity_test.go:
```diff
@@ -0,0 +1,47 @@
+		})
+	})
+
+	t.Run("Duration()", func(t *testing.T) {
```
Would have methods always has `verbs` and not `nouns`

mlimaloureiro commented on repositories/tasks/deleter_repository.go:
```diff
@@ -0,0 +1,8 @@
+package tasks // import github.com/mlimaloureiro/golog/repositories/tasks
+
+// DeleterTaskRepositoryInterface is a TaskRepositoryInterface that can also delete tasks directly
+type DeleterTaskRepositoryInterface interface {
```
what was the reason to have a dedicated interface?

NMFR replied:
Some repositories might have implementations for this 2 operations (for example a SQL repository). This interface would allow TaskService to call those methods directly instead of using GetTasks() -> SetTasks() to delete tasks. Something similar to StarterPauserTaskRepositoryInterface.

But currently no repositories implement it.

Should i remove it?

mlimaloureiro commented on repositories/tasks/file/csv/repository_test.go:
```diff
@@ -0,0 +1,219 @@
+	"time"
+
+	tasksModel "github.com/mlimaloureiro/golog/models/tasks"
+	tasksRepositories "github.com/mlimaloureiro/golog/repositories/tasks"
```
why is it plural ?

mlimaloureiro commented on repositories/tasks/file/csv/repository_test.go:
```diff
@@ -0,0 +1,219 @@
+func getFileContent(t *testing.T, filePath string) string {
+	stringBytes, err := ioutil.ReadFile(filePath)
+	assert.NoError(t, err)
+	return string(stringBytes)
```
always add a new line in the last return, let's also add it to our styleguide

NMFR replied:
Do you mean like this:
``` go
func getFileContent(t *testing.T, filePath string) string {
	stringBytes, err := ioutil.ReadFile(filePath)
	assert.NoError(t, err)
	
	return string(stringBytes)
}
````
?

When there is only one line this should be ignored right?
``` go
func GetPi() float64 {
    return 3.14
}
```

mlimaloureiro commented on repositories/tasks/memory/repository.go:
```diff
@@ -0,0 +1,67 @@
+
+	if taskPtr := repository.tasks.GetByIdentifier(task.Identifier); taskPtr == nil {
+		repository.tasks = append(repository.tasks, taskClone)
+	} else {
```
always avoid if/elses

NMFR committed:
Fix codestyle issues
